docs/adr/adr-009-telemetry.md

```diff
@@ -0,0 +1,226 @@
+may happen during the operation.
+
+Spans also form a parent tree, meaning that each span associated to a process can have multiple sub processes or child
+spans and vise-versa. Altogether, this feature allows to see the whole trace of execution of any part of the system, no 
+matter how complex it is. This is exactly what we need to analyze our reconstruction performance.
+
+An example of a root span in `ipld.Retriever`:
+```go
+import "go.opentelemetry.io/otel"
+
+var tracer = otel.Tracer("share")
+
+func (r *Retriever) Retrieve(ctx context.Context, dah *da.DataAvailabilityHeader) (*rsmt2d.ExtendedDataSquare, error) {
+  ctx, span := tracer.Start(ctx, "retrieve-square")
+  defer span.End()
+
+  span.SetAttributes(
+    attribute.Int("size", len(dah.RowsRoots)),
+    attribute.String("data_hash", hex.EncodeToString(dah.Hash())),
+  )
+  ...
+}
+```
+
+The child span in `ipld.Retriever.Reconstruct`:
+```go
+	ctx, span := tracer.Start(ctx, "reconstruct-square")
+	defer span.End()
+
+	// and try to repair with what we have
+	err := rs.squareImported.Repair(rs.dah.RowsRoots, rs.dah.ColumnRoots, rs.codec, rs.treeFn)
+	if err != nil {
+		span.RecordError(err)
+		return nil, err
+	}
+```
+
+And the quadrant request event:
+```go
+    span.AddEvent("requesting quadrant", trace.WithAttributes(
+        attribute.Int("axis", q.source),
+        attribute.Int("x", q.x),
+        attribute.Int("y", q.y),
+        attribute.Int("size", len(q.roots)),
+    ))
+```
+> The above is only examples related to our code and is a subject to change.
+
+#### Backends connection 
+
+Jaeger example
+```go
+    // Create the Jaeger exporter
+    exp, err := jaeger.New(jaeger.WithCollectorEndpoint(jaeger.WithEndpoint(url)))
+    if err != nil {
+        return nil, err
+    }
+	// then the tracer provider
+    tp := tracesdk.NewTracerProvider(
+        // Always be sure to batch in production.
+        tracesdk.WithBatcher(exp),
+        // Record information about this application in a Resource.
+        tracesdk.WithResource(resource.NewWithAttributes(
+            semconv.SchemaURL,
+            semconv.ServiceNameKey.String(service),
+            attribute.String("environment", environment),
+            attribute.Int64("ID", id),
+        )),
+    )
+	// and set it globally to be used across packages
+    otel.SetTracerProvider(tp)
+	
+	// then close it elsewhere
+    tp.Shutdown(ctx)
+```
+
+### Metrics
+
+// WIP
+
+## Considerations
+
+* Tracing performance
+  * _Every_ method calling two more functions making network request can affect overall performance
+
+## Other
+
+As you will see in the examples below, tracing looks similar to logging and have almost the same semantics. In fact,
+tracing is debug logging on steroids, and we can potentially consider dropping conventional _debug_ logging once we 
+fully cover our codebases with the tracing. Same as logging, traces can be pipe out into the stdout as prettyprinted
+event log.
+
```

Comment:
I think additionally to the proposed Glossary a set of "further reading" links or "related links" would be good. E.g. the opentelemetry docs themselves are really great and the semantics of [metrics](https://opentelemetry.io/docs/reference/specification/metrics/semantic_conventions/) and [tracing](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/) are useful. Similarly the prometheus docs on metric naming could be useful for devs integrating metrics into node: https://prometheus.io/docs/practices/naming/#metric-names or the intro docs to prometheus: https://prometheus.io/docs/introduction/overview/

Reply:
Good point. WIll add it

Reply:
done